### PR TITLE
Implementation of ink_env 2.0

### PR DESCRIPTION
**Pull request by Robbepop**
This implements https://github.com/paritytech/ink/issues/187.
Note that throughout the process of this PR the proposal of #187 is highly outdated.
This PR reflects the latest design state of #187.

Compared to the old environment implementation this offers interfaces to operate more efficiently with the SRML contracts module using less allocations. Also it does a much better job of abstracting away the actual interface between host (SRML contracts) and client (the contract).

### CODE REVIEW

File: core/src/env2/traits.rs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    ) where
+        O: scale::Output + AsRef<[u8]> + Reset,
+        R: scale::Encode;
+
+    /// Returns a random hash given the subject.
+    fn random<I>(buffer: &mut I, subject: &[u8]) -> Self::Hash
+    where
+        I: AsMut<[u8]> + EnlargeTo;
+
+    /// Prints the contents as a single line.
```
Comment (ascjones) [inline core/src/env2/traits.rs]: Perhaps add a comment that this should not be used in "production"

File: core/src/env2/call.rs [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    fn encode_to<T: scale::Output>(&self, dest: &mut T) {
+        dest.write(self.bytes.as_slice());
+    }
+}
+
+impl scale::Decode for CallData {
+    fn decode<I: scale::Input>(
+        input: &mut I,
+    ) -> core::result::Result<Self, scale::Error> {
+        let mut bytes = Vec::new();
```
[the remark this answers is not in the document]
Comment (Robbepop) [reply core/src/env2/call.rs]: Yep, that could potentially reduce the number of allocations required. :+1:

File: core/src/env2/call.rs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+        }
+        Ok(Self { bytes })
+    }
+}
+
+/// Represents a return type.
+///
+/// Used as a marker type to differentiate at compile-time between invoke and evaluate.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct ReturnType<T>(PhantomData<T>);
```
[the remark this answers is not in the document]
Comment (Robbepop) [reply core/src/env2/call.rs]: Just forgot to do so. Thanks.

File: core/src/env2/test/accessor.rs [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+    GasPrice => |instance| {
+        instance.borrow().state.gas_price.to_origin()
+    }
+
+    GasLeft => |instance| {
+        instance.borrow().exec_context.gas_left.to_origin()
+    }
+
+    NowInMs => |instance| {
+        instance.borrow().exec_context.gas_left.to_origin()
```
Comment (ascjones) [inline core/src/env2/test/accessor.rs]: Copy/paste error: should be `now_in_ms`

**Pull request merged**